theano/gpuarray/opt.py

```diff
@@ -1592,6 +1593,77 @@ def local_gpua_lift_abstractconv_graph(op, context_name, inputs, outputs):
 
 
 @register_opt()
+@op_lifter([pool.Pool])
+@register_opt2([pool.Pool])
+def local_gpu_pool(op, ctx_name, inputs, outputs):
+    assert op.__props__ == ('ignore_border', 'mode', 'ndim')
+    inp, ws, stride, pad = inputs
+    nd = op.ndim
+    if nd not in (2, 3):
+        return
+    inp = gpu_contiguous(as_gpuarray_variable(inp, ctx_name))
+
+    op = GpuPool(op.ignore_border, op.mode, op.ndim)
+    if inp.ndim == nd + 2:
+        return op(inp, ws, stride, pad)
+    else:
+        # reshape to 4D or 5D with 2 non-pooling dimensions
+        inp_padded = pad_dims(inp, 2, nd)
+        ret_padded = op(inp_padded, ws, stride, pad)
+        return unpad_dims(ret_padded, inp, 2, nd)
```

Comment:
Tests for 4d and 5d inputs were passing. My commit which tried to fix conflicts with cudnn optimizations broke them (cudnn optimization applied before new pooling optimizations). For now, I can remove optimizations and tests with `ignore_border = True`, so there are will be no conflict with cudnn optimizations.